### PR TITLE
feat(AureliaUx): Use event for two ways binding

### LINKED ISSUE #140: Ux components & user input subscription
At the moment, we declared a `value` field with `@bindable({ defaultBindingMode: bindingMode.twoWay })` for user input (`ux-input`, `ux-textarea`, `ux-select` [i'm doing this]). I think we should follow the way we subscribe to built-in inputs: dispatch `change` event to notify subscribers. To do this, `Ux-core` needs to provide a method to intercept `determineDefaultBindingMode`  at [here](https://github.com/aurelia/templating-binding/blob/master/src/syntax-interpreter.js#L40) for subscribing to change event instead of binding directly to view model field.

This enable more fine-grained control on how to notify subscribers, benefits: `ux-select`, at least. I feel like declaring `@bindable({ defaultBindingMode: bindingMode.twoWay })` should be declared on application level components, not component level lib, plus the async nature of `bindable` will be problematic in the future I expect.

What do you think @EisenbergEffect @jdanyow @ZHollingshead

### PR DESCRIPTION
**Pull request by bigopon**
Changes:

### core

- requires https://github.com/aurelia/binding/pull/656 for observer typings
- adds `booleanAttr` binding behavior to get / set attribute & aria for boolean props
- takes dependency on `aurelia-templating-binding` (now intercepts `SyntaxInterpreter` class)
- adds new configuration API:

  ```ts
    registerUxElementConfig(config: {
      tagName: string
      properties: {
        [propertyName: string]: {
          defaultBindingMode: bindingMode,
          getObserver(
            obj: Element,
            propertyName: string,
            observerLocator: ObserverLocator
          ): InternalPropertyObserver
        }
      }
    });
  ```

    Example:

  ```js
    ux.registerUxElementConfig({
      tagName: 'ux-checkbox',
      properties: {
        checked: {
          defaultBindingMode: bindingMode.twoWay,
          getObserver(element, propertyName, observerLocator) {
            return new CheckedObserver(element, uxCheckboxChangeHandler, observerLocator);
          }
        }
      }
    });

    const uxCheckboxChangeHandler = {
      subscribe(target, callbackOrListener) {
        target.addEventListener('change', callbackOrListener, false);

        return function() {
          target.removeEventListener('change', callbackOrListener, false);
        };
      }
    };
  ```

----

#### overall goal for `<ux-input/>`, `<ux-textarea/>` and probably `<ux-select/>`, `<ux-checkbox/>`, `<ux-radio/>`:

-  will all notify their subscribers via `change` event.
- will not support `containerless`, as it relies on element dispatching event to notify subscribers. Can support aliasing via `as-element`.
- will mimic built-in controls by defining getter/setter on their elements, pointing to underlying view model values

#### `<ux-input/>`

- `value` is now real number when type is `number`. Thus empty string, input will give value `null`. `NaN` is also coerced into `null`.
- `null` / `undefined` are better handled, as they are kept as is instead of coerced into empty string.
- generally values as `null`/ `undefined` are kept as is, converted to number if type `number`, `string` if other types.

#### `<ux-checkbox/>`

- behaves in the same way with native `<input type="checkbox"/>`, add `indeterminate`

#### `<ux-textarea/>`

- generally values as `null`/ `undefined` are kept as is, converted `string` otherwise

---------

## TODO

- [x] Cleanup `<ux-checkbox/>`
- [x] Refactor `<ux-radio />`
  - [x] Add indeterminate
- [x] Refactor `<ux-textarea />`

----

* resolves #140

### CODE REVIEW

**Comment by bigopon** [comment]
This is for the reason I mentioned in #140

**Comment by bigopon** [comment]
@ZHollingshead I refactor `<ux-input/>` with the changes, please pull and give it a shot. 

`<ux-input/>` now has `rawValue: string` to handle internal value and `processRawValue` to handle `type = number`

**Comment by bigopon** [comment]
@ZHollingshead @EisenbergEffect I updated the PR with some high level design decision. Would be nice if you could give some feedback based on new input

File: packages/checkbox/src/index.ts [hunk 1]
```diff
@@ -1,9 +1,33 @@
...
 export { UxCheckboxTheme } from './ux-checkbox-theme';
 
 export function configure(config: FrameworkConfiguration) {
+  config.container.get(AureliaUX).registerUxElementConfig(uxCheckBoxConfig);
   config.globalResources([
     PLATFORM.moduleName('@aurelia-ux/checkbox/ux-checkbox')
   ]);
 }
+
+const uxCheckBoxConfig = {
```
Comment (ZHollingshead) [inline packages/checkbox/src/index.ts]: We need to be sure to extend this configuration to the [`@aurelia-ux/components`](https://github.com/aurelia/ux/tree/master/packages/components) package.

File: packages/input/src/index.ts [hunk 2]
```diff
@@ -1,9 +1,33 @@
...
 export { UxInputTheme } from './ux-input-theme';
 
 export function configure(config: FrameworkConfiguration) {
+  config.container.get(AureliaUX).registerUxElementConfig(uxInputConfig);
   config.globalResources([
     PLATFORM.moduleName('@aurelia-ux/input/ux-input')
   ]);
 }
+
+const uxInputConfig = {
```
Comment (ZHollingshead) [inline packages/input/src/index.ts]: We need to be sure to extend this configuration to the [`@aurelia-ux/components`](https://github.com/aurelia/ux/tree/master/packages/components) package.

File: packages/radio/src/index.ts [hunk 3]
```diff
@@ -1,9 +1,33 @@
...
 export { UxRadioTheme } from './ux-radio-theme';
 
 export function configure(config: FrameworkConfiguration) {
+  config.container.get(AureliaUX).registerUxElementConfig(uxCheckBoxConfig);
   config.globalResources([
     PLATFORM.moduleName('@aurelia-ux/radio/ux-radio')
   ]);
 }
+
+const uxCheckBoxConfig = {
```
Comment (ZHollingshead) [inline packages/radio/src/index.ts]: We need to be sure to extend this configuration to the [`@aurelia-ux/components`](https://github.com/aurelia/ux/tree/master/packages/components) package.

Comment (ZHollingshead) [reply packages/radio/src/index.ts]: Shouldn't this be `uxRadioConfig` rather than `uxCheckBoxConfig`?

File: packages/switch/src/index.ts [hunk 4]
```diff
@@ -1,9 +1,33 @@
...
 export { UxSwitchTheme } from './ux-switch-theme';
 
 export function configure(config: FrameworkConfiguration) {
+  config.container.get(AureliaUX).registerUxElementConfig(uxCheckBoxConfig);
   config.globalResources([
     PLATFORM.moduleName('@aurelia-ux/switch/ux-switch')
   ]);
 }
+
+const uxCheckBoxConfig = {
```
Comment (ZHollingshead) [inline packages/switch/src/index.ts]: We need to be sure to extend this configuration to the [`@aurelia-ux/components`](https://github.com/aurelia/ux/tree/master/packages/components) package.

Comment (ZHollingshead) [reply packages/switch/src/index.ts]: Shouldn't this be `uxSwitchConfig` instead?

File: packages/textarea/src/index.ts [hunk 5]
```diff
@@ -1,9 +1,33 @@
...
 export { UxTextareaTheme } from './ux-textarea-theme';
 
 export function configure(config: FrameworkConfiguration) {
+  config.container.get(AureliaUX).registerUxElementConfig(uxTextAreaConfig);
   config.globalResources([
     PLATFORM.moduleName('@aurelia-ux/textarea/ux-textarea')
   ]);
 }
+
+const uxTextAreaConfig = {
```
Comment (ZHollingshead) [inline packages/textarea/src/index.ts]: We need to be sure to extend this configuration to the [`@aurelia-ux/components`](https://github.com/aurelia/ux/tree/master/packages/components) package.

File: packages/radio/src/index.ts [hunk 6]
```diff
@@ -1,9 +1,33 @@
...
+    checked: {
+      defaultBindingMode: bindingMode.twoWay,
+      getObserver(element: Element, _: string, observerLocator: ObserverLocator) {
+        return new CheckedObserver(element, uxCheckboxChangeHandler, observerLocator);
+      }
+    }
+  }
+};
+
+const uxCheckboxChangeHandler = {
```
Comment (ZHollingshead) [inline packages/radio/src/index.ts]: Shoudn't this be `uxRadioChangeHandler`?

File: packages/radio/src/ux-radio.ts [hunk 7]
```diff
@@ -1,144 +1,132 @@
...
+  PaperRipple,
+  normalizeBooleanAttribute,
+  linkProperty
+} from '@aurelia-ux/core';
 import { UxRadioTheme } from './ux-radio-theme';
+import { ElementEvents, DOM } from 'aurelia-framework';
 
 const theme = new UxRadioTheme();
 
+export interface UxRadioElement extends HTMLElement {
```
Comment (ZHollingshead) [inline packages/radio/src/ux-radio.ts]: Should we export this in the index file?

File: packages/switch/src/ux-switch.ts [hunk 8]
```diff
@@ -1,19 +1,31 @@
...
+import { computedFrom, observable } from 'aurelia-binding';
 import { inject } from 'aurelia-dependency-injection';
-import { StyleEngine, UxComponent, PaperRipple, normalizeBooleanAttribute } from '@aurelia-ux/core';
+import { StyleEngine, UxComponent, PaperRipple, normalizeBooleanAttribute, linkProperty } from '@aurelia-ux/core';
 import { UxSwitchTheme } from './ux-switch-theme';
+import { DOM, ElementEvents } from 'aurelia-framework';
 
 const theme = new UxSwitchTheme();
 
+export interface UxSwitchElement extends HTMLElement {
```
Comment (ZHollingshead) [inline packages/switch/src/ux-switch.ts]: Should we export this in the index file?

**Comment by bigopon** [comment]
@ZHollingshead All above done. The last 2 commits:

- Exports more in each component `index.ts`
- Exports more in `@aurelia-ux/components/index.ts`
- Renames `Textarea` to `TextArea` (please revert / override if it was intentional)
- Add temporary exports to `aurelia-binding` while waiting for aurelia/binding#656

For `@aurelia-ux/components/index`, calling configuration should do the job, maybe we don't have to do anything in there.

**Comment by ZHollingshead** [comment]
This looks good. Will wait to merge, there is a fix going out for the `aurelia-ux/core` package and then we can merge this in.

**Pull request merged**